src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs

```diff
@@ -19,88 +17,112 @@ namespace Microsoft.CodeAnalysis;
 
 internal sealed class LinkedFileDiffMergingSession(Solution oldSolution, Solution newSolution, SolutionChanges solutionChanges)
 {
-    internal async Task<LinkedFileMergeSessionResult> MergeDiffsAsync(IMergeConflictHandler mergeConflictHandler, CancellationToken cancellationToken)
+    internal async Task<LinkedFileMergeSessionResult> MergeDiffsAsync(IMergeConflictHandler? mergeConflictHandler, CancellationToken cancellationToken)
     {
         var sessionInfo = new LinkedFileDiffMergingSessionInfo();
 
-        var linkedDocumentGroupsWithChanges = solutionChanges
-            .GetProjectChanges()
-            .SelectMany(p => p.GetChangedDocuments())
-            .GroupBy(d => oldSolution.GetDocument(d).FilePath, StringComparer.OrdinalIgnoreCase);
+        using var _ = PooledDictionary<string, List<(Document newDocument, ImmutableArray<byte> newContentHash)>>.GetInstance(out var filePathToNewDocumentsAndHashes);
+        foreach (var documentId in solutionChanges.GetProjectChanges().SelectMany(p => p.GetChangedDocuments()))
+        {
+            var newDocument = newSolution.GetRequiredDocument(documentId);
+            var filePath = newDocument.FilePath;
+            Contract.ThrowIfNull(filePath);
 
-        var linkedFileMergeResults = new List<LinkedFileMergeResult>();
+            if (!filePathToNewDocumentsAndHashes.TryGetValue(filePath, out var newDocumentsAndHashes))
+            {
+                newDocumentsAndHashes = [];
+                filePathToNewDocumentsAndHashes.Add(filePath, newDocumentsAndHashes);
+            }
+
+            var newText = await newDocument.GetValueTextAsync(cancellationToken).ConfigureAwait(false);
+            var newContentHash = newText.GetContentHash();
+            // Ignore any linked documents that we have the same contents as.  
+            if (newDocumentsAndHashes.Any(t => t.newContentHash.SequenceEqual(newContentHash)))
+                continue;
```

Comment:
this is the meat of the change.  we just ignore siblings with the same hash as one we've already seen.  that means it has the same text, so no point trying to fetch that text (and subtexts) to figure out the diff/merge info.